### ISSUE TITLE
Comprehensive Kernel Benchmarking Suite for Performance & VRAM Profiling

### ISSUE THREAD

**Issue by tzervas** [tzervas/unsloth-rs#2]
## Overview

Develop a comprehensive benchmarking and profiling suite to validate the performance targets for all GPU kernels implemented in unsloth-rs.

## Scope

Benchmark all fused kernels across hardware configurations:
- **Flash Attention** (Q·K^T·V single-pass)
- **RoPE** (Rotary Position Embeddings)
- **RMSNorm** (with optional bias)
- **SwiGLU** (fused activation)
- **Memory Pool** efficiency
- **Gradient Checkpointing** memory impact

### 1. Benchmark Infrastructure (`benches/kernels.rs`)

```rust
use criterion::{criterion_group, criterion_main, Criterion, BenchmarkId};
use unsloth_rs::kernels:: *;

fn benchmark_flash_attention(c: &mut Criterion) {
    let mut group = c.benchmark_group("flash_attention");
    
    for seq_len in [512, 1024, 2048, 4096] {
        group.bench_with_input(
            BenchmarkId::new("cuda", seq_len),
            &seq_len,
            |b, &s| {
                // CubeCL kernel benchmark
            },
        );
        
        group.bench_with_input(
            BenchmarkId::new("cpu", seq_len),
            &s,
            |b, &s| {
                // CPU reference benchmark
            },
        );
    }
    
    group.finish();
}
```

### 2. VRAM Profiling

- Document memory usage per kernel configuration
- Track peak allocation vs baseline PyTorch/naive implementation
- Validate 70-80% VRAM reduction target
- Profile memory pool effectiveness

### 3. Hardware Configurations

**Primary Targets:**
- NVIDIA A100 (80GB)
- NVIDIA RTX 4090 (24GB)
- NVIDIA RTX 3090 (24GB)

**Configurations to Test:**
- Batch sizes: 1, 4, 8, 16
- Sequence lengths: 512, 1024, 2048, 4096, 8192
- Hidden sizes: 768, 1024, 2048, 4096
- Model architectures: LLaMA-7B, LLaMA-13B equivalent configs

### 4. Performance Metrics

**Speed:**
- ✅ Minimum 2x speedup vs naive implementation
- Target:  2-5x speedup on attention-heavy workloads
- Measure throughput (tokens/sec) and latency (ms per forward pass)

**Memory:**
- ✅ 70-80% VRAM reduction with gradient checkpointing
- Document memory breakdown per kernel
- Compare against candle-core baseline operations

**Accuracy:**
- ✅ Numerical equivalence within 1e-5 tolerance
- Validate CubeCL GPU vs CPU reference outputs
- Test edge cases (batch=1, very long sequences)

### 5. Benchmark Outputs

Generate comprehensive reports:

```markdown
## Flash Attention Performance (A100)

| Config | Naive (ms) | Fused (ms) | Speedup | VRAM Naive (GB) | VRAM Fused (GB) | Reduction |
|--------|-----------|-----------|---------|-----------------|-----------------|----------|
| B=4, S=2048, H=4096 | 45.2 | 18.7 | 2.4x | 12.3 | 3.8 | 69% |
| B=8, S=4096, H=4096 | 182.5 | 51.3 | 3.6x | 48.7 | 12.1 | 75% |
```

### 6. Integration with CI/CD

- Run benchmarks on PR to `experimental` branch
- Flag performance regressions automatically
- Publish results to GitHub Pages or repo docs

## Acceptance Criteria

- [ ] All kernels benchmarked across target hardware configurations
- [ ] Benchmarks demonstrate **2-5x speedup** for fused operations
- [ ] VRAM profiling confirms **70-80% reduction** with checkpointing enabled
- [ ] Numerical equivalence validated (1e-5 tolerance) for all kernels
- [ ] Results documented in `BENCHMARKS.md` or similar in repo root
- [ ] Benchmark suite integrated into `benches/kernels.rs` with Criterion
- [ ] Performance regression testing enabled in CI workflow
- [ ] GPU occupancy metrics collected (target >50%)

## Related Issues

- Depends on:  #[flash-attention], #[rope], #[rmsnorm], #[swiglu]
- Milestone: **Phase 5: Benchmarking & Validation**

## References

- [Unsloth Performance Claims](https://github.com/unslothai/unsloth)
- [CubeCL Benchmarking Docs](https://github.com/tracel-ai/cubecl)
- [Flash Attention Paper](https://arxiv.org/abs/2205.14135) - Performance baselines
- Kernel Optimization Skill:  `.github/skills/unsloth-kernel-optimization/SKILL.md`

**Comment by tzervas** [comment]
CPU benches exist. GPU benches that ignore D2H/H2D are not product evidence (DEBT.md / README). Keep this open for **honest** GPU benches after #87 P0d. Do not publish 2x/70% from kernel-only microbenchmarks.
